File: app/src/main/java/fr/gstraymond/db/json/JsonDeck.java
package fr.gstraymond.db.json;


import android.content.Context;

import com.magic.card.search.commons.json.MapperUtil;
import com.magic.card.search.commons.log.Log;
import com.squareup.moshi.Moshi;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.util.ArrayList;
import java.util.List;

import fr.gstraymond.models.CardWithOccurrence;

public class JsonDeck {

    private Log log = new Log(this);

    private Context context;
    private MapperUtil<List<CardWithOccurrence>> mapperUtil;

    public JsonDeck(Context context, Moshi objectMapper) {
        this.context = context;
        this.mapperUtil = MapperUtil.fromCollectionType(objectMapper, CardWithOccurrence.class);
    }

    public List<CardWithOccurrence> load(String deckId) {
        try {
            FileInputStream inputStream = context.openFileInput("deck_" + deckId);
            return new ArrayList<>(mapperUtil.read(inputStream));
        } catch (FileNotFoundException e) {
            log.w("get: %s", e);
            return new ArrayList<>();
        }
    }

    public void save(String deckId, List<CardWithOccurrence> cards) {
        try {
            FileOutputStream fos = context.openFileOutput("deck_" + deckId, Context.MODE_PRIVATE);
            fos.write(mapperUtil.asJsonString(cards).getBytes());
            fos.close();
        } catch (Exception e) {
            log.e("save", e);
        }
    }

    // change occurrence
}

File: app/src/main/kotlin/fr/gstraymond/db/json/JsonDeck.kt
package fr.gstraymond.db.json

import com.magic.card.search.commons.json.MapperUtil
import com.squareup.moshi.Moshi
import fr.gstraymond.android.CustomApplication
import fr.gstraymond.models.CardWithOccurrence
import fr.gstraymond.utils.getId

class JsonDeck(customApplication: CustomApplication,
               moshi: Moshi,
               deckId: String) : JsonList<CardWithOccurrence>(
        customApplication,
        MapperUtil.fromCollectionType(moshi, CardWithOccurrence::class.java),
        deckId) {
    override fun CardWithOccurrence.uid() = card.getId()
}

class JsonDeckBuilder(private val customApplication: CustomApplication,
                      private val moshi: Moshi) {
    fun build(deckId: Int): JsonDeck = JsonDeck(customApplication, moshi, "$deckId")
}
